google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,384 @@
+   *     through {@code WriteStream}
+   * @return Builder
+   */
+  public static Builder newBuilder(String streamName, Table.TableSchema tableSchema) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    return new Builder(streamName, tableSchema, null);
+  }
+
+  /**
+   * newBuilder that constructs a JsonStreamWriter builder.
+   *
+   * @param streamName name of the stream that must follow
+   *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+   * @param tableSchema The schema of the table when the stream was created, which is passed back
+   *     through {@code WriteStream}
+   * @param client
+   * @return Builder
+   */
+  public static Builder newBuilder(
+      String streamName, Table.TableSchema tableSchema, BigQueryWriteClient client) {
+    Preconditions.checkNotNull(streamName, "StreamName is null.");
+    Preconditions.checkNotNull(tableSchema, "TableSchema is null.");
+    Preconditions.checkNotNull(client, "BigQuery client is null.");
+    return new Builder(streamName, tableSchema, client);
+  }
+
+  /** Closes the underlying StreamWriter. */
+  @Override
+  public void close() {
+    this.streamWriter.close();
+  }
+
+  public static final class Builder {
+    private String streamName;
+    private BigQueryWriteClient client;
+    private Table.TableSchema tableSchema;
+
+    private TransportChannelProvider channelProvider;
+    private CredentialsProvider credentialsProvider;
+    private BatchingSettings batchingSettings;
+    private RetrySettings retrySettings;
+    private ExecutorProvider executorProvider;
+    private String endpoint;
+    private OnSchemaUpdateRunnable onSchemaUpdateRunnable;
+
+    /**
+     * Constructor for JsonStreamWriter's Builder
+     *
+     * @param streamName name of the stream that must follow
+     *     "projects/[^/]+/datasets/[^/]+/tables/[^/]+/streams/[^/]+"
+     * @param tableSchema schema used to convert Json to proto messages.
+     * @param client
+     */
+    private Builder(String streamName, Table.TableSchema tableSchema, BigQueryWriteClient client) {
+      this.streamName = streamName;
+      this.tableSchema = tableSchema;
+      this.client = client;
+    }
+
+    /**
+     * Setter for the underlying StreamWriter's TransportChannelProvider.
+     *
+     * @param channelProvider
+     * @return Builder
+     */
+    public Builder setChannelProvider(TransportChannelProvider channelProvider) {
+      this.channelProvider =
+          Preconditions.checkNotNull(channelProvider, "ChannelProvider is null.");
+      return this;
+    }
+
+    /**
+     * Setter for the underlying StreamWriter's CredentialsProvider.
+     *
+     * @param credentialsProvider
+     * @return Builder
+     */
+    public Builder setCredentialsProvider(CredentialsProvider credentialsProvider) {
```

Comment:
are these necessary (CredentialsProvider, BatchingSettings, setRetrySettings....)? If they are, should we move them to constructor to avoid the case that user forgot to set them?

Reply:
Yes these settings are necessary cause they can customize the StreamWriter's function. And so here I'm trying to let users set the settings in the JsonStreamWriter Builder class instead of the JsonStreamWriter class because it seems to make more sense having the Builder class to handle all settings. If the user's don't set them, there won't be any exceptions and the StreamWriter will be created by default settings.